File: android/app/src/main/java/sol/vortexx/android/savedgifs/impl/HttpSavedGifs.kt
package sol.vortexx.android.savedgifs.impl

import io.ktor.client.call.body
import io.ktor.client.request.*
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.http.isSuccess
import kotlinx.serialization.Serializable
import sol.vortexx.android.net.impl.VortexHttpClient
import sol.vortexx.android.savedgifs.api.SavedGif
import sol.vortexx.android.savedgifs.api.SavedGifs
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class HttpSavedGifs @Inject constructor(
    private val http: VortexHttpClient,
) : SavedGifs {
    @Serializable private data class Wrap(val gifs: List<SavedGif>)
    @Serializable private data class AddBody(val url: String, val width: Int, val height: Int)

    override suspend fun list(): List<SavedGif> = runCatching {
        val resp = http.client.get("api/saved_gifs")
        if (!resp.status.isSuccess()) return@runCatching emptyList()
        resp.body<Wrap>().gifs
    }.getOrDefault(emptyList())

    override suspend fun add(url: String, width: Int, height: Int): SavedGif? = runCatching {
        val resp = http.client.post("api/saved_gifs") {
            contentType(ContentType.Application.Json)
            setBody(AddBody(url, width, height))
        }
        if (!resp.status.isSuccess()) return@runCatching null
        resp.body<SavedGif>()
    }.getOrNull()

    override suspend fun remove(id: Long) {
        runCatching { http.client.delete("api/saved_gifs/$id") }
    }
}

File: ios/Modules/Sources/SavedGifs/impl/HttpSavedGifs.swift
import Foundation
import Net

public actor HttpSavedGifs: SavedGifs {
    private let http: HttpClient
    public init(http: HttpClient) { self.http = http }

    private struct Wrap: Codable { let gifs: [SavedGif] }
    private struct AddBody: Codable { let url: String; let width: Int; let height: Int }

    public func list() async -> [SavedGif] {
        let req = HttpRequest.get("/api/saved_gifs")
        return (try? await http.send(req, Wrap.self).gifs) ?? []
    }

    public func add(url: String, width: Int, height: Int) async -> SavedGif? {
        guard let req = try? HttpRequest.postJson("/api/saved_gifs", body: AddBody(url: url, width: width, height: height)) else { return nil }
        return try? await http.send(req, SavedGif.self)
    }

    public func remove(id: Int64) async {
        let req = HttpRequest.delete("/api/saved_gifs/\(id)")
        _ = try? await http.send(req)
    }
}
